Feature/deregister exemption

timstone opened the pull request:
New back office tests to deregister an exemption and then deregister the last exemption to cause the whole registration to be deregistered.

timstone committed:
improved choosing exemption steps so selecting multiple exemptions is possible

timstone committed:
wip

timstone committed:
added deregister exemption steps

timstone committed:
wip

timstone committed:
set default exemption selection

timstone committed:
completed deregistering exemption steps

timstone committed:
improved feature details

Cruikshanks commented on features/backoffice_deregister_exemption.feature:
```diff
@@ -0,0 +1,21 @@
+Feature: Back office user deregisters a waste exemption activity from a registration
+  As a back office user
+  I want to deregister a waste exemption
+  So that the exemption is deregistered for a valid registration with mutlipe exemptions
```
There's a spelling error (mutlipe = multiple) but this is more a case of could we give a better reason, and one that is more generic. We shouldn't be referencing multiple exemptions and instead just have separate scenarios which covers registrations with single and multiple exemptions (these can come later).

timstone replied:
I think the point is that it there are more than one exemptions then it's only that exemption that is deregistered. Happy to change the wording though.

Cruikshanks commented on features/backoffice_deregister_exemption.feature:
```diff
@@ -0,0 +1,21 @@
+
+  @happy_path
+  Scenario: Deregistration of an exemption for an active registration by a NCCC user
+  
```
Delete the line between the Scenario and the steps.

Cruikshanks commented on features/backoffice_deregister_exemption.feature:
```diff
@@ -0,0 +1,21 @@
+    And I complete a registration
+
+  @happy_path
+  Scenario: Deregistration of an exemption for an active registration by a NCCC user
```
I think we can give more meaning to this feature if we refer specifically to ceasing and/or revoking an exemption. So the steps would read

```gherkin
When I [cease] an exemption
Then I can see the exemption status is [Ceased]
And when I [cease] the last active exemption
Then I can see the registration itself is [Ceased]
```

The things I've put in brackets could be passed as arguments to the steps, so they would still be reusable, but using the actual terms rather than the generic *deregistered/deregister* reads better and gives us more scope for testing different scenarios later.

timstone replied:
Agreed, I was planning on doing that but then got distracted...

Cruikshanks commented on features/page_objects/deregistration_page.rb:
```diff
@@ -3,21 +3,35 @@ class DeregistrationPage < SitePrism::Page
   section(:nav_bar, AdminNavBarSection, ".add-bottom-margin .container")
 
-  def submit(args = {})
+  def deregister_registration(args = {})
```
It looks like we're wrapping both the exemption deregistration page and the registration deregistration page into one page object. These should be split into distinct page objects.

Also I'd stick with the pattern of having a `submit()` method into which we pass in the values we wish to enter/select on a page from the step, and then let that method handle what needs to be done, rather than naming it specific to the page/feature.

Cruikshanks commented on features/page_objects/registration_details_page.rb:
```diff
@@ -1,9 +1,14 @@
+  element(:registration_status, :xpath, ".//*[@id='registration']/div/table/tbody/tr[3]/td[2]/span")
   element(:deregister, "#deregister-enrollment")
-
+  element(:exemption_details_section, ".row:nth-child(3) .col-xs-11")
```
Could we work together to move the different sections of the page into actual SitePrism sections? Appreciate in this case it won't be reused by another page, but I think it will help break up the logic and make it clear what belongs to where and what logic is relevant.

Cruikshanks commented on features/step_definitions/back_office/edit_registration_steps.rb:
```diff
@@ -9,9 +9,9 @@
   @app.registration_details_page.deregister.click
 
-  @app.deregistration_page.submit(
+  @app.deregistration_page.deregister_registration(
```
Already mentioned about leaving the method name as `submit()`.

Cruikshanks commented on features/step_definitions/back_office/edit_registration_steps.rb:
```diff
@@ -26,16 +26,56 @@
   @app.registration_details_page.deregister.click
 
-  @app.deregistration_page.submit(
+  @app.deregistration_page.deregister_registration(
```
Already mentioned about leaving the method name as `submit()`.

Cruikshanks commented on features/step_definitions/back_office/edit_registration_steps.rb:
```diff
@@ -26,16 +26,56 @@
+  @app.search_page.search_button.click
+
+  @app.search_page.first_search_result.click
+  sleep(10)
```
Need to remove the `sleep()`.

Cruikshanks commented on features/step_definitions/back_office/edit_registration_steps.rb:
```diff
@@ -26,16 +26,56 @@
+  @app.search_page.first_search_result.click
+  @app.registration_details_page.exemption_details_section.click
+  @app.registration_details_page.deregister_last_exemption.click
+  @app.deregistration_page.deregister_exemption(
```
Already mentioned about leaving the method name as `submit()`.

Cruikshanks commented on features/page_objects/choose_exemptions_page.rb:
```diff
@@ -10,18 +10,36 @@ class ChooseExemptionsPage < SitePrism::Page
+  # Identifies what tab that exemption is on using the identify_tab method
+  # Selects that tab using the select_tab method
+  # Then clicks on the relevant checkbox for that exemption number
+  def submit(args = { exemptions: %w(S1 D2) })
```
I'd say don't use a default value for the `args` object. I think when it comes to maintaining the project in the future it'll be clearer if you can see in the step what exemptions you're selecting, rather than just see `@app.choose_exemptions_page.submit` with no args.

Appreciate it might mean the same thing is repeated in multiple places, but I'd prefer seeing the arguments rather than wondering what has been done behind the scenes.

timstone replied:
ok don't have a strong view on having a default and so happy to revert it back to how it was previously. I think I was just concerned that one test relies on having two exemptions so that they can be deregistered in sequence to then cause a registration deregistration.

Cruikshanks reviewed (changes requested):
Some changes to be done. If you like I'd be happy to make them whilst you get on with some FRAE stuff tomorrow.

timstone commented:
If you'd like to polish up this work then I'm happy for you to progress it, I'll start looking at FRAE, thanks.

Cruikshanks committed:
Refactored deregistration features, steps, & pages

This change builds on what came previously, but refactors the original features, steps, and pages in the following way.

- Multiple deregistrations features refactored to two; deregister exemption and deregister registration
- Single deregistration page refactored to separate deregister exemption and registration pages
- Regustration details refactored to a new section from registration page
- Exemption details refactored to a new section from registration page
- Site details refactored to a new section from registration page
- Deregistration steps moved to a new `deregister_steps.rb` file
- `farm_questions_page.rb` refactored to work with arguments rather than set choices
- Sections now include a constant called `SELECTOR` (where appropriate) which should be used when adding a section to a page

Also included are some housekeeping and helper changes

- Sign out from back office after each scenario now handled by an `after scenario` hook
- Ability to override back office url via an environment variable
- Details about how the back office url is determined added to the README
- `rubocop.yml` updated to increase maximum method length

Cruikshanks commented:
I've just got a couple of things to check off, and I also need to run the full suite (turns out I have trouble connecting to WEX from home!) but essentially I'm done. Feel free to start having a look through and we can discuss in more detail when I get in @timstone .

Cruikshanks committed:
Fix failing site_addr steps / beautify new reg

Cruikshanks committed:
First step in renaming site_details

Cruikshanks committed:
Complete site details rename

Cruikshanks reviewed (approved):
Obviously I'm happy with this now but you need to check for yourself @timstone . Regards the naming of **SiteDetails** I've gone with **LocationDetails** to get around the issue discussed, and will also raise an error.

When you come to merging this in, please use the squash option in order to clean up our commit message
